core/embed/trezorhal/entropy.c

```diff
@@ -0,0 +1,74 @@
+/*
+ * This file is part of the Trezor project, https://trezor.io/
+ *
+ * Copyright (c) SatoshiLabs
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "entropy.h"
+
+#include <string.h>
+
+#include "common.h"
+#include "flash.h"
+#include "rand.h"
+
+#include "stm32f4xx_ll_utils.h"
+
+extern __IO uint32_t uwTick;
+
+uint8_t HW_ENTROPY_DATA[HW_ENTROPY_LEN];
+uint8_t SW_ENTROPY_DATA[SW_ENTROPY_LEN] = {0};
+static size_t sw_entropy_index = {0};
+
+void collect_hw_entropy(void) {
```

Comment:
> ...it just reads out secrets into a buffer before the areas containing those secrets get locked out. 

Yes, I was aware of that.

> Putting it into this file together with the sw_entropy stuff reinforces this false notion.

I don't think it does. Anyway, the reason why I moved it from `common.c` is that I think `common.c` should contain only the functions that are used in many other modules (which is not true for `collect_hw_entropy`) and `collect_hw_entropy` makes from that file a heap of functions that don't have its own place.

I like `collect_hw_secrets` more than `collect_hw_entropy`. What about putting it into a separate file, for example `hardware_secrets.c`.

Reply:
I am in favor of renaming to `collect_hw_secrets()`. The function doesn't belong in this file, because it's not gathering any entropy. Now that I look at it again, it should be a static function in embed/firmware/main.c, because that's the only place where it's called.

Reply:
I don't like moving it into `main.c`. I would rather leave `main.c` as simple as possible.

What about putting it into a separate file, for example `hardware_secrets.c`.

Reply:
I am unable to see the rationale behind creating two new files for a 23-line helper function that is used in only one place, but more importantly https://wiki.sei.cmu.edu/confluence/display/c/DCL19-C.+Minimize+the+scope+of+variables+and+functions.